Philip docs (#2)
NLnetLabs/connectbyname: notes/dns-client-proxy.md

A local proxy may provide some benefit to short-lived applications by
caching results. In particular if the system uses a so called 'public
DNS resolver'. In general we assume that the cache is tagged according
to the source of a reply and the transport it is received on.

This allows
applications to communicate using Do53 and still get the privacy
benefits from using more secure protocols over the internet. However,
such a setup leaves the application with no control over which transport
the proxy uses. This document introduces EDNS(0) options that allow a
stub resolver to request certain transports and allow the proxy to report
capabilities and actual transports that are available.

With respect to DNSSEC, we assume that an application that needs
DNSSEC validation, for example, for DANE validation or SSHFP, will
perform the DNSSEC validation within the application itself and does not
trust the proxy. The proxy can of course do DNSSEC validation as well.
Important however, is that an untrusted proxy cannot provide an application
with a traditional (unsigned) trust anchor.

For the transport configuration we expect three levels of details. The

> first is a choice between anythings goes, unauthenticated encryption and
> authenticated encryption. The second level is where the application also

Author: Replace "anything goes" with something indication that things are tried in order perhaps? (not sure)

```suggestion
first is a choice between requiring authenticated encryption, also allowing unauthenticated encryption or doing opportunistic encryption on an best effort basis. The second level is where the application also
```

specifies the names and/or IP addresses of upstream resolvers. The
third level is where the application also specifies which transports
(Do53, DoT, DoH, DoQ) are allowed to be used. A final transport parameter
is the outgoing interface that is to be used.

For authentication we can have a mix of PKI and DANE. Options are one of
the two and not the other, both or one of the two.

A poorly constructed DNS client proxy may forward DNS packets without

> properly handling of EDNS(0) options. To detect and prevent this,

Author: Yes, I agree that that would be poorly constructed, however there is a RFC for DNS proxies which has ambivalent statements about this when trying to take into account the transparency principle for DNS Proxies. Have a read of section 3 of that RFC: 

https://www.rfc-editor.org/rfc/rfc5625.html#section-3

I guess it's a balance. Perhaps we should replace "poorly constructed" with something a bit less judgmental?

Reviewer: I improved the wording with a reference to RFC 5625 that such proxies are expected to exist. I also added an example.

      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
      |     Interface Name Length     |                               |
      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+                               |
      ~                 Interface Name                                ~
      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
~~~

where

{newline="true"}
OPTION-CODE

: To be decided

OPTION-LENGTH

: Lenght of this option excluding the OPTION-CODE and OPTION-LENGTH fields

U

: force the use of unencrypted communication (Do53)

UA

: require unauthenticated encryption

A

: require authenticated encryption

P

: authenticate using a PKI certificate

D

: authenticate using DANE

DO

: disallow other transports (transports that are not explicitly listed)

A53,AT,AH2,AH3,AQ

: allow respectively Do53, DoT, DoH H2, DoH H3, DoQ

D53,DT,DH2,DH3,DQ

: disallow respectively Do53, DoT, DoH H2, DoH H3, DoQ

Z

: reserved, MUST be zero when sending, MUST be ignored when received

Addr Type

: Type of addresses, IPv4 or IPv6

Addr Length

: length of the addresses in octets. Must be a multiple of 4 for IPv4 and

> a multiple of 16 for IPv6

Author: Do we need to provision for scoped IPv6 addresses? I.e. like: fe80::1ff:fe23:4567:890a%eth2
I think we should somehow...

Reviewer: We should make mention of this in the text, but my idea is that the interface field can be used to store for that

Reviewer: I added text that for IPv6 link-local addresses the zone ID can be put in the interface field.

IPv4 or IPv6-address(es)

: list of IPv4 or IPv6 addresses

Domain Name Length

: length of Domain Name. Zero if there is no Domain Name

Domain Name

: domain name for authentication or resolving IP addresses

SvcParams Length

: length of SvcParams

SvcParams

: Service parameters

Interface Name Length

: length of Interface Name

Interface Name

: name of outgoing interface for transport connections

This option is designed to give control over what level of detail it
wants to specify. The first 5 flags (U, UA, A, P, and D) give generate
requirements for properties of DNS transports that are used by the
client proxy. The U flag specifies no encryption, the UA flag, at
least unauthenticated encryption and the A flag require authenticated
encryptions. With the U, UA, and A flags clear, an effort is made to
reach authenticated encryption, if that fails unauthenticated encryption

> and if that fails, fall back to an unencrypted transport.

Author: Perhaps an overview with all the possible combinations would make things a bit clearer:

0. `U` = 0, `UA` = 0, `A` = 0: Try `A`, then `UA`, then `U`
1. `U` = 0, `UA` = 0, `A` = 1: Require `A`
2. `U` = 0, `UA` = 1, `A` = 0: Require `UA` without trying `A`
3. `U` = 0, `UA` = 1, `A` = 1: Try `A`. If `A` failed require `UA`
4. `U` = 1, `UA` = 0, `A` = 0: Do `U` without trying `A` and `UA`
5. `U` = 1, `UA` = 0, `A` = 1: What would this mean? Same as 0?
6. `U` = 1, `UA` = 1, `A` = 0: Try `UA`. If `UA` failed do `U`.
7. `U` = 0, `UA` = 0, `A` = 0: Same as 0: Try `A`, then `UA`, then `U`

Reviewer: I think these flags are mutually exclusive. Is there a combination of falgs that makes sense?

The P and D flags allow the application to require
a specific authentication mechanism (PKI or DANE). When both flags are set,

> PKI and DANE are required together. If no flags are set, are least one of

Author: DANE has 4 usage types (see https://tools.ietf.org/id/draft-ietf-dane-ops-02.html#rfc.section.2 ): PKIX-TA, PKIX-EE, DANE-TA and DANE-EE. The last two do not require PKIX validation. Does `P` and `D` bit both set mean that PKIX has to be valid even if usage types DANE-TA or DANE-EE are present?

Perhaps it would be good to enlist all possibilities with text what it means. Something like this?

1. `P` and `D` bot unset: default operation (DNSSEC is optional)
    - Fall back to PKIX if DANE lookup is non-existant or insecure or indeterminate
    - Authentication fails if DANE RRs are BOGUS.
    - DANE usage types PKIX-TA and PKIX-EE require PKIX validation.
2. Only `D` set:  Same as 1, but DNSSEC is required (i.e. no indeterminate status allowed)
    - Fall back to PKIX if DANE lookup is non-existant or insecure
    - Authentication fails if DANE RRs are BOGUS.
    - DANE usage types PKIX-TA and PKIX-EE require PKIX validation.
3. Only `P` set: Do only PKIX authentication without trying and/or doing lookups for DANE
4. Both `P` and `D` set: Same as 2?

In all cases a Certificate Authority repository is requied to perform PKIX validation.

Reviewer: My idea about 'D' is that a secure DANE record has to be present otherwise it is a failure. Do you think the semantics should be different?

the two has to succeed if authenticated encryption is required.

The next flags provide more detailed control over which transports
should be used or not. For each of 5 different transports
(Do53, DoT, DoH with ALPN h2, DoH with ALPN h3, DoQ) there is a flag
to allow (A53,AT,AH2,AH3,AQ) or disallow (D53,DT,DH2,DH3,DQ) the use of the
transport. There is space to add more transports later.

To future proof applications, there is a single flag DO, that disallows
transports that are not explicitly listed. With this flag clear,
the application allows future transports. With the flag set, the
application has to explicitly list which transports can be used.
For example, by setting only DO and AT, the application forces the

> use of DoT.

Author: I don't think I understand how this would work precisely. Does below make sense:

1. `DO` = 0:
    - **all** transports are in the pool of potentially usable transports
    - `D53`, `DT`, `DH2`, `DH3` and `DQ` remove those transports from the pool
2. `DO` = 1:
    - **no** transports are in the pool of potentially usable transports
    - `A53`, `AT`, `AH2`, `AH3` and `AQ` add those transports to the pool

What would conflicting `A*` and `D*` otherwise mean?

Reviewer: This summary of how DO work is nice. I'll add it.

Reviewer: Done

Finally, an application can specify its own resolvers or rely on the
resolvers that are know to the proxy. If ADN Length and Addr Length
are both zero, then the application requests to resolvers known to
the proxy. [Note: it is unclear at the moment what to do with any
Service Parameters]

If the application specifies only an authentication-domain-name then the
proxy is expected to resolve the name to addresses. If only addresses
are specified then the proxy assumes that no name is known (though
a PKI certificate may include an address literal in the subjectAltName).
If both a name and address are specified then the proxy will use the
specified addresses and use the name for authentication.

To simplify the encoding of the option, an option with addresses will
have either IPv4 or IPv6 addresses. If the application wants to specify

> both IPv4 and IPv6 address for a certain authentication-domain-name

Author: ```suggestion
both IPv4 and IPv6 addresses for a certain authentication-domain-name
```

Reviewer: Done

then it has to include two options.

When present, Service Parameters specify how to connect. Otherwise it is

> up to the proxy to try various possibilities.

Author: Service Parameters can also contain IPv4 and IPv6 addresses (hints) are they simply added to the list of addresses to try with the policy?
Service Parameters can also contain an alpn parameter containing a list of alpns. Does that mean that only the specified transports are available on the in the same Service Parameters given addresses and port?
I guess we need to brain storm a bit about this.

Reviewer: https://datatracker.ietf.org/doc/html/draft-ietf-add-dnr-07 says: "The service parameters MUST NOT include "ipv4hint" or "ipv6hint" SvcParams as they are superseded by the included IP addresses."

There is something to be said for that. Certainly to keep the protocol simple.

I think ALPNs in Service Parameters are complex. We certainly need to give a summary of how that works.

> 

Author: Description of the Interface option. Can there be more than one? Because we also have the length of the option, there could in principle be more than one.

Reviewer: My reasoning is that because you already need separate options for v4 and v6 addresses in this encoding, it should not be a big deal to have an option for each outgoing interface. Unless there are existing RFCs where a host ends up with a collection of interfaces (from service discovery, policies, etc).

returns both an ANCHORS-XML and an ANCHORS-P7S, then the application verifies
the trust anchor using the trust anchor certificate (which needs to come
with the application).

# Server Processing

Proxies are encouraged to cache options that appear in requests under the
assumption that a stub resolver will send multiple requests. If a proxy
caches DNS responses then the proxy MUST tag cached responses with the
properties of the DNS transport. When responding to later requests,
the proxy returns a cached entry only if the parameters of the DNS transport
match what is specified in the request.

When a proxy receives a new set of requirements, the proxy compiles a list of
addresses to connect to and a list of transports to try per address.
The proxy SHOULD prefer more private transports over less private ones.

If the proxy cannot obtain a connection to a recursive resolver in a way that
matches the provided policy, then the proxy sets the BADPROXYPOLICY
response code in the reply. In addition the proxy SHOULD try to find at least
one connection to a recursive resolver. If the proxy did not find an
alternative connection to a recursive resolver (also if it didn't try) then

> the proxy includes a PROXY POLICY Option with all flags set to zero,

Author: ```suggestion
the proxy includes a PROXY CONTROL Option with all flags set to zero,
```

Reviewer: Done

resolvers then the proxy generates options with the properties of those
connection, with a maximum of three connections.

If the proxy finds a PROXY SCOPE Option, then it calculates the scope from
the source address. If the scope is larger then specified in the option,
then the proxy returns a BADPROXYPOLICY response code. The proxy sets the value of
Scope to the actual scope of the source address of the request.

If the request contains a TRUST ANCHOR Option, then the proxy tries to
fetch the trust anchor XML and p7s files if it does not have them already.
If fetching one or both fails then the proxy sets the corresponding
length to zero. It is not clear how long the proxy can cache this information.
[@RFC7958] Does not describe how long these documents can be cache.
A simple solution is to take the Expires header in the HTTP reply.

# Connection Between Stub Resolver And Proxy

Absent other configuration, a stub resolver that implements this standard
SHOULD connect to the proxy using Do53 and as remote address either ::1 or
127.0.0.1. In particular, the stub resolver SHOULD avoid using name
servers listed in files such as /etc/resolv.conf.

> There reason for this is to simplify the integration of local DNS proxies

Author: ```suggestion
The reason for this is to simplify the integration of local DNS proxies
```

Reviewer: Done

in existing environments. If the stub resolver ignores /etc/resolv.conf
then the proxy can use that information to connect to recursive resolvers.

If no DNS server is responding to queries sent using Do53 to ::1 and 127.0.0.1,
or if the response indicates that this standard is not supported, then
the stub resolver MAY fall back to traditional configuration methods, such
as /etc/resolv.conf. However, in that case the stub resolver MUST make
sure that doing so does not violate the policy set by the application.

# Security Considerations

# IANA Considerations

IANA has assigned the following DNS EDNS0 option codes:

     Value   Name           Status     Reference
    ------- -------------- ---------- -----------

>      TBD     PROXY POLICY   Standard   RFC xxxx

Author: ```suggestion
     TBD     PROXY CONTROL   Standard   RFC xxxx
```

Reviewer: Done

%%%
title = "Control Options For DNS Client Proxies"
abbrev = "codcp"
area = "Internet"
workgroup = "ADD"

[seriesInfo]
status = "standard"
name = "Internet-Draft"
value = "draft-homburg-codcp-00"
stream = "IETF"

date = 2022-05-02T00:00:00Z

[[author]]
initials="P.C."
surname="Homburg"
fullname="Philip Homburg"
organisation = "NLnet Labs"
  [author.address]
  email = "philip@nlnetlabs.nl"
%%%

.# Abstract

The introduction of many new transport protocols for DNS in recent years
(DoT, DoH, DoQ)
significantly increases the complexity of DNS stub
resolvers that want to support these protocols. A practical way forward
is to have a DNS client proxy in the host operating system. This allows
applications to communicate using Do53 and still get the privacy
benefit from using more secure protocols over the internet. However,
such a setup leaves the application with no control over which transport
the proxy uses. This document introduces EDNS(0) options that allow a
stub resolver to request certain transport and allow the proxy to report
capabilities and actual transports that are available.

{mainmatter}

# Definitions

Do53

> : The original, plain text DNS transport as described in [@RFC1035].

Author: ```suggestion
: The original, plain text DNS transport as described in [@RFC1034;@RFC1035].
```

Reviewer: done

Typically, UDP is used, with the DNS server listening on port 53. Sometimes,
for example, for large responses, TCP is used, also on port 53.

DoH

: DNS over HTTPS as described in [@RFC8484].

DoT

: DNS over TLS as described in [@RFC7858]

DoQ

: DNS over QUIC ([@RFC9000]) as described in [I-D.ietf-dprive-dnsoquic],
not to be confused with DNS over HTTP/3 which also uses QUIC

EDNS(0) Option

: An option as described in [@RFC6891]

h2

: This TLS ALPN identifies HTTP/2 as described in [@RFC7540]

h3

: This TLS ALPN identifies HTTP/3, which is HTTP over QUIC and

> is described in I.D.ietf-quic-http (expired draft)

Author: ```suggestion
is described in [@RFC9114]
```

Reviewer: RFC9114 is HTTP/3 not DNS over QUIC

~~~

where

{newline="true"}
OPTION-CODE

: To be decided

OPTION-LENGTH

: Length of this option excluding the OPTION-CODE and OPTION-LENGTH fields

U

: force the use of unencrypted communication (Do53)

UA

: require unauthenticated encryption

A

: require authenticated encryption

P

: authenticate using a PKIX certificate

D

: authenticate using DANE

DD

: by default disallow other transports
(transports that are not explicitly listed)

A53,AT,AH2,AH3,AQ

: allow respectively Do53, DoT, DoH H2, DoH H3, DoQ

D53,DT,DH2,DH3,DQ

: disallow respectively Do53, DoT, DoH H2, DoH H3, DoQ

Z

: reserved, MUST be zero when sending, MUST be ignored when received

Addr Type

: Type of addresses, The value 0 if no addresses are included, the value 1 for
IPv4, and the value 2 for IPv6.

Addr Length

: length of the addresses in octets. Must be a multiple of 4 for IPv4 and
a multiple of 16 for IPv6. This field can be zero if no addresses are
specified.

IPv4 or IPv6-address(es)

: list of IPv4 or IPv6 addresses

Domain Name Length

: length of Domain Name. Zero if there is no Domain Name

Domain Name

: domain name for authentication or resolving IP addresses. The domain name

> is encoding in uncompressed DNS wire format.

Author: ```suggestion
is encoded in uncompressed DNS wire format.
```

Reviewer: done

flag. The proxy SHOULD return a BADPROXYPOLICY error if either or both of the 
P or D flags is set and the A flag is clear.

The next flags provide more detailed control over which transports
should be used or not. For each of 5 different transports
(Do53, DoT, DoH with ALPN h2, DoH with ALPN h3, DoQ) there is a flag
to allow (A53,AT,AH2,AH3,AQ) or disallow (D53,DT,DH2,DH3,DQ) the use of the
transport. There is space to add more transports later. Note that setting
the A flags and the D flag for a protocol (for example, setting both the
A53 and the D53 flags) is not allowed and a proxy SHOULD reject such a request.

To future proof applications, there is a single flag DD, that by default
disallows transports that are not explicitly listed. With this flag clear,
the application allows all transports that are not explicitly disallowed
(including future transports). With the flag set, the
application has to explicitly list which transports can be used.
For example, by setting only DD and AT, the application forces the
use of DoT.

When DD = 0:

* all transports are in the pool of potentially usable transports
* D53, DT, DH2, DH3 and DQ remove those transports from the pool

> 

Author: ```suggestion
* The values of A53, AT, AH2, AH3 and AQ are ignored

```

Reviewer: done

Reviewer: Hmm, ignored is wrong, because earlier there is 'Note that setting
the A flag and the D flag for a protocol (for example, setting both the
A53 and the D53 flags) is not allowed and a proxy SHOULD reject such a request.'

I'll change ignored
 to irrelevant

When DD = 1:

* no transports are in the pool of potentially usable transports
* A53, AT, AH2, AH3 and AQ add those transports to the pool

> 

Author: ```suggestion
* The values of D53, DT, DH2, DH3 and DQ are ignored

```

Reviewer: Also irrelevant instated of ignored

through an IPv6 link-local address.
IPv6 link-local addresses are special in that they require a zone to be
specified, either explicitly or implicitly. Typically for a link-local
address that appears as a source or destination address,
the zone is implicitly the zone of the link the packet travels on.
For packets that travel between hosts, there is no goed way to explictly
specify the zone of a link-local address because two different hosts
do not agree on zone names. However, if the proxy is on the same host
as the application, then the zone identifier for the link-local address
can be specified in the Interface field. For this purpose an interface
name can also be an interface index expressed as a decimal string.

When present, Service Parameters specify how to connect. Otherwise it is
up to the proxy to try various possibilities. For Service Parameters,
the values of the ipv4hint and ipv6hint fields are ignored. Addresses
can only be specified using the addresses field in the PROXY CONTROL Option.

Associated with this option is a new error, BADPROXYPOLICY. When
a proxy cannot meet the requirements in a PROXY CONTROL Option or the

> option is malformed, it returns this error.

Author: Before this also contained text that stated that the error was returned " along with a PROXY CONTROL option that lists what it can do.". Do we no longer want that? I think it is said before that the PROXY CONTROL option can also be used to report on capabilities, and also to probe for capabilities.

Reviewer: Yes, I removed it. I think it is too confusing to use such a value in an application. If it has a lot of diagnostic value, then we could make it optional for the proxy to return this.

Just trying to keep it simple and remove mandated complexity.

# PROXY SCOPE OPTION

~~~
      +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
   0: |                          OPTION-CODE                          |
      +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
   2: |                         OPTION-LENGTH                         |
      +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
   4: |                          Scope                                |
      +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
~~~

{newline="true"}
OPTION-CODE

: To be decided

OPTION-LENGTH

: Length of this option excluding the OPTION-CODE and OPTION-LENGTH fields

Scope

: Scope of the source address of a request. Scope can have the following
values:

     Value | Scope
    -------|-------
     0     | Undefined
     1     | Host local
     2     | Link local
     3     | Site local
     4     | Global

The purpose of this option is to deal with proxies that forward
DNS traffic without first removing any EDNS(0) options. The option requests
the DNS proxy that processes the option to report the scope of the source

> address. 

Author: I see you removed the text "The application can specify that the connection between the stub resolver and the proxy should have, for example, at most host local scope.". Can clients no longer restrict the scope?

Reviewer: Yes. I decided that application can check, but the proxy does not enforce. That simplifies things because the proxy just includes the option with the right value, no action needs to be taken by the proxy.

should be initialized according to the needs of the application.
In addition the PROXY SCOPE Option can be added. In requests, the Scope
field is set to undefined.

If the stub resolver receives a reply without a PROXY CONTROL Option
included in the reply, then stub resolver has to assume that traffic will
have Do53 levels of privacy.  Similarly, a lack of a PROXY SCOPE Option
implies a global scope.

If the stub resolver receives a BADPROXYPOLICY error then the proxy was
unable to meet the requirements of the PROXY CONTROL Option.

### Probing

In cases where the stub resolver expects a local DNS proxy, or where the
stub resolver has (a limited) fall back to more private transports, or
when the security policy of the application is such that is better to fail
than send queries over Do53, the stub resolver first sends a probing
query to verify that the proxy supports the PROXY CONTROL and
PROXY SCOPE Options.

This request queries "resolver.arpa" for SOA records.
The proxy MUST implement this as a
Special Use Domain Name. The actual response is not important. The
important part is that the proxy returns PROXY CONTROL and PROXY SCOPE
Options as described in this document or sets

> the response code to BADPROXYPOLICY if it cannot meet specified policy.

Author: "or sets the response code" implies that in that case no PROXY CONTROL option is returned, but did we not initially have the idea that it would return such an option filled with the capabilities.
Also perhaps if quering for "resolver.arpa" SOA one could send an empty PROXY CONTROL option to just ask for a capability report?

Reviewer: I added the following at the end of the PROXY CONTROL OPTION section:
If the proxy returns a BADPROXYPOLICY error, the proxy MAY include
a PROXY CONTROL Option that lists what the proxy can do. For example,
if authenticated encryption is not possible, but unauthenticated is,
then the proxy may include an option that has the UA bit set.
